src/Lykke.Job.CandlesHistoryWriter.Services/HistoryMigration/CandlesFiltrationService.cs

```diff
@@ -0,0 +1,242 @@
+                {
+                    dateFrom = candle.Timestamp;
+                    dateTo = candle.LastUpdateTimestamp.TruncateTo(smallerInterval)
+                        .AddIntervalTicks(1, smallerInterval);
+
+                    var smallerCandles = await _candlesHistoryRepository.GetCandlesAsync(candle.AssetPairId,
+                        smallerInterval, priceType, dateFrom, dateTo);
+
+                    // Trying to reconstruct the candle from the corresponfing smaller interval candles, if any.
+                    ICandle updatedCandle = null;
+                    foreach (var smallerCandle in smallerCandles)
+                    {
+                        if (updatedCandle == null)
+                            updatedCandle = Candle
+                                .Copy(smallerCandle)
+                                .RebaseToInterval(interval);
+                        else
+                            updatedCandle = updatedCandle
+                                .ExtendBy(smallerCandle
+                                    .RebaseToInterval(interval));
+                    }
+
+                    // If the candle is not empty (e.g., there are some smaller interval candles for its construction),
+                    // we should update it in storage. Otherwise, it is to be deleted from there.
+                    if (updatedCandle != null)
+                        currentCandlesToReplace.Add(updatedCandle);
+                    else
+                        currentCandlesToDelete.Add(candle);
+                }
+
+                if (currentCandlesToDelete.Count > 0)
+                {
+                    countDeleted = await _candlesHistoryRepository.DeleteCandlesAsync(currentCandlesToDelete);
+                    deletedCount += countDeleted;
+                }
+
+                if (currentCandlesToReplace.Count > 0)
+                {
+                    var countReplaced = await _candlesHistoryRepository.ReplaceCandlesAsync(currentCandlesToReplace);
+                    replacedCount += countReplaced;
+                }
+            }
+
+            return (deletedCandlesCount: deletedCount, replacedCandlesCount: replacedCount);
+        }
+
+        #region "Private"
+
+        // Calculating the earliest and the latest dates for the biggest interval candles fetching
+        private async Task<(DateTime dateFrom, DateTime dateTo)> GetDateTimeRangAsynce(string assetPairId, CandlePriceType priceType)
+        {
+            var firstBiggestCandle =
+                await _candlesHistoryRepository.TryGetFirstCandleAsync(assetPairId,
+                    Constants.StoredIntervals.Last(),
+                    priceType);
+
+            // If we have no such a candle in repository, we return fake dates for further decision making in the caller code
+            if (firstBiggestCandle == null)
+                return (dateFrom: DateTime.MinValue, dateTo: DateTime.MinValue);
```

Comment:
Be careful here. Not all external systems can correctly handle date 1.1.0001(0)

Reply:
Ok, but this is a private utility method. It is not used outside the class.